### PR TITLE
Remove gmaps4rails gem [fixes #134262481]

### PR DESCRIPTION
**Pull request by Maroo-b**
Addresses: https://www.pivotaltracker.com/story/show/134262481

### CODE REVIEW

**Comment by tansaku** [comment]
thanks for this @Maroo-b - we can just delete the old implementation maybe? we'll have the history of it in git right?

Also I'm thinking that hound is just making these things unmanageable, maybe we should switch it off ...

**Comment by Maroo-b** [comment]
@tansaku sure thing, for the moment the missing tasks are : 
- to implement the handle event for closing the infobox
- add some refactoring for the new js file.
- fix hound comments

By the way I noted this dynamic setting for map height, [source](https://github.com/AgileVentures/LocalSupport/blob/develop/app/assets/javascripts/google_map.js.coffee#L33), do you remember why it was added ? because I seems that it doesn't have any effect on the current version.

**Comment by tansaku** [comment]
no idea about dynamic height setting - @johnnymo87 might know ...

**Comment by Maroo-b** [comment]
@tansaku I have re implemented the map functionalities without gmaps4rails gem but I have failing jasmine tests to fix.

**Comment by Maroo-b** [comment]
@tansaku and @johnnymo87 can you please assist to fix this javascript [test](https://github.com/AgileVentures/LocalSupport/blob/develop/spec/javascripts/google_map_spec.js) ?

File: app/assets/javascripts/google_maps/custom_marker.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    panes.overlayImage.appendChild(div);
+  }
+
+  var point = this.getProjection().fromLatLngToDivPixel(this.latlng);
+
+  // By default the the top left corner of custom marker is used to set marker position
+  // check the explancation from: http://humaan.com/custom-html-markers-google-maps/
+  if (point) {
+    div.style.left = (point.x - ($('.custom-marker').width() / 2)) + 'px';
+    div.style.top = (point.y - $('.custom-marker').height()) + 'px';
```
Comment (johnnymo87) [inline app/assets/javascripts/google_maps/custom_marker.js]: Do these markers have a consistent height and width? If so, I wonder if we could avoid doing this calculation for each marker?

File: app/assets/javascripts/maps.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  var ib = new InfoBox();
+  var ibOptions = {
+    pixelOffset: new google.maps.Size(-151, 10)
+  };
+
+  var boxText = document.createElement("div");
+  boxText.setAttribute("class", "arrow_box")
+  ibOptions.content = boxText;
+
+  for (i = 0; i < markerData.length; i++) {
```
Comment (johnnymo87) [inline app/assets/javascripts/maps.js]: How do you feel about using `forEach` here?

**Comment by johnnymo87** [comment]
@Maroo-b Testing this javascript may be difficult due to it's dependence on global variables, e.g. `google`, `document`, `this.setMap`, etc. I see some options:
1. Refactor to make all dependencies explicitly injected.
2. Load the real google map libraries in the test suite.
3. Load a mock of the google map libraries in the test suite.

I've briefly explored these options, here's what I've seen so far:

1. This works until `InfoBox`. That library depends on `google` and `document` being globally available.
2. Add `- "//maps.google.com/maps/api/js?v=3&amp;libraries=geometry<%= ENV['GMAP_API_KEY'].nil? ? '' : "&key=#{ENV['GMAP_API_KEY']}" %>"` to `src_files` in `spec/javascripts/support/jasmine.yml`. I think the tests will need to wait for everything to load.
3. This works until `this.setMap`. This function and others seem to magically arrive with the real google library. I'm not familiar enough with javascript to understand how they magically get mixed into the `CustomMarker` class.

**Comment by Maroo-b** [comment]
@johnnymo87 thanks for your feedback, I think the easiest solution is to extract marker_reader to a separate function but is it really worth it to be tested ? It seems it's an implementation test not behavior one no ?

**Comment by johnnymo87** [comment]
Yeah, I agree it's an implementation test, but if I am not sure if there is any other way to assert that markers are drawn on the provided coordinates.

**Comment by Maroo-b** [comment]
From the test file ([source](https://github.com/AgileVentures/LocalSupport/blob/develop/spec/javascripts/google_map_spec.js)), it looks like we are testing if function `marker_data` is able to read data from 'data-markers' attributes.

Regarding to check if markers are drawn I checked that the functionality is verified by those cucumber tests ( [maps](https://github.com/AgileVentures/LocalSupport/blob/develop/features/home_page/map.feature), [volunteer ops](https://github.com/AgileVentures/LocalSupport/blob/develop/features/volunteer_opportunities/map_volunteer_opportunities.feature)) 

for the moment I'll delete that file to make the test suite pass.

File: app/assets/javascripts/maps.js (hunk 2 above)
Comment (Maroo-b) [inline app/assets/javascripts/maps.js]: thx for the note I'll make an update.

**Comment by johnnymo87** [comment]
Yeah, we probably have a good deal of coverage via selenium. I'm not sure if it's 100%, though. For instance, what if we swapped the latitude and longitude of the coordinates ... would any cukes fail? If not, we probably need lower level js tests.

**Comment by Maroo-b** [comment]
@johnnymo87 good point, 

When I was working on this PR I used the default google markers to check if the custom markers are correctly positioned, maybe we can think how to implement this test with jasmine. 

Regarding swapping lat and lng inputs to google library, technically it will draws the marker to another position that why I suggested to compare  default markers with custom ones.

**Comment by tansaku** [comment]
@Maroo-b many thanks for working on this - great to see that old coffee script thrown out - I'm much more comfortable with pure JS.

I notice that we now have a custom markers_builder.rb class - is that not what the gem was providing for us before?  If memory serves we are removing the gem to make tighter list/map integration possible is that right?

We're making some custom ruby code here to give us more flexibility about how we use the map, is that right?

**Comment by Maroo-b** [comment]
@tansaku thank you for checking the PR.

I extracted the markers_builder.rb from gmaps4rails gem with it's spec file it's used to create the options for the markers.

The main change as you noted is in the javascript part, know we are using the native google js library so it's more easier to customize and check google maps docs for reference. For example I added the option to  center a maker on click [source](https://github.com/AgileVentures/LocalSupport/pull/397/files?diff=unified#diff-5c1d378ee5df16a03145c365cb754616R53) which is now so easy and straightforward.

**Comment by tansaku** [comment]
@Maroo-b great! so we are good to go?

What do you think of these two code climate issues? do you think we can safely ignore them?  @johnnymo87 ?

Could we ultimately be extracting our js code into an npm module that we could share with others?

Is the gmap4rails gem not being maintained?  Should we be doing PRs against that?

File: app/services/gmaps/markers_builder.rb [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    class MarkerBuilder
+
+      attr_reader :object, :hash
+      def initialize(object)
+        @object = object
+        @hash   = {}
+      end
+
+      def call(&block)
```
Comment (tansaku) [inline app/services/gmaps/markers_builder.rb]: maybe prefer `run` here? http://blog.silvabox.com/musings-on-service-objects/

**Comment by tansaku** [comment]
also, do we need some jasmine tests on custom_marker.js and maps.js ?

**Comment by tansaku** [comment]
also, can we find some way to get test coverage information for jasmine?

File: app/services/gmaps/markers_builder.rb (hunk 3 above)
Comment (Maroo-b) [inline app/services/gmaps/markers_builder.rb]: @tansaku this file was extracted from Gmaps4rails gem, the only method I added was `generate` in class `MarkersBuilder` [source](https://github.com/AgileVentures/LocalSupport/pull/397/files#diff-044234d81e24db3a723cd82292b9758dR35) 
The `MarkerBuilder` instance accepts a block that's why it seems that naming the method to `call` here was chosen to reveal the intention of calling blocks.

**Comment by Maroo-b** [comment]
@tansaku for a custom google maps integration, it seems easier to use the native google maps js library. And now with pure js it's simple to check customization of event handling and managing opening infobox based on list items.

So for jasmine tests, I find that maps integration is well covered with integration tests but it will be amazing if we could add tests for `custom_marker.js`  because when I was working on this PR I used the default google markers to check if the custom markers are correctly positioned.

File: app/assets/javascripts/maps.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        boxText.innerHTML = item.infowindow;
+        $(ib.content_).find('.close').click(function(){
+          ib.close();
+        });
+        ib.open(map, marker);
+        map.panTo(ib.getPosition());
+
+      }
+
+    })(marker, item));
```
Comment (johnnymo87) [inline app/assets/javascripts/maps.js]: Code climate [noticed](https://codeclimate.com/github/AgileVentures/LocalSupport/pull/397) that this function is invoked when passed to the `addListener` function. It's interesting that this invoked function returns a function. @Maroo-b, does `marker` and `item` still work in the inner-most function without being enclosed in this invoked function?

**Comment by johnnymo87** [comment]
@Maroo-b Would you mind tagging all the relevant cucumber tests with an `@maps` tag?

I agree, the native google maps library is too easy to integrate with to merit making this a public library.

I think that javascript could use some testing. It will be challenging, though. The downside of the easy integration is that it's _too_ easy. `google` is a global variable, and `CustomMarker` inherits from `google.maps.OverlayView`, introducing a lot of magic functionality. You can't reliably test it without either (1) loading the real google maps libraries in the test suite, or (2) writing an extremely thin wrapper around it, depend on that instead, and test against that.

That said, this otherwise looks like a very nice implementation, so if the cucumber tests are sufficient, I say we ship it and open a separate ticket for the javascript testing.

@tansaku I believe the `gmaps4rails` gem is still maintained, but all we're doing here is taking one file from it and tossing the rest, which probably wouldn't be welcomed upstream.

**Comment by Maroo-b** [comment]
@johnnymo87 I update the PR to add `maps` tags to cucumber tests.

**Comment by tansaku** [comment]
@Maroo-b @johnnymo87 that's great - thanks for the explanation, I'll make a ticket for the javascript unit testing ... https://www.pivotaltracker.com/story/show/136036569

**Comment by tansaku** [comment]
I'm just going to run this locally to check for anything we might have missed and then get it merged in

**Pull request merged**